lib/CRMS/RightsPredictor.pm

```diff
@@ -0,0 +1,140 @@
+# Used in Commonwealth project UI for populating and updating UI with
+# rights based on metadata and researcher-provided data.
+# Also used for New Year rights rollover.
+
+# This could be extended to non-Commonwealth countries.
+
+# TODO: this code relies heavily on publication/copyright date,
+# and as a result makes Bib API calls multiple times in the course
+# of a single review. Extending the Metadata module to use a local
+# database cache for volumes in queue would enhance responsiveness
+# an eliminate some unnecessary "paranoia" code.
+
+use strict;
+use warnings;
+use utf8;
+
+use Carp;
+use POSIX;
+
+sub new {
+  my ($class, %args) = @_;
+  my $self = bless {}, $class;
+  $self->{$_} = $args{$_} for keys %args;
+  Carp::confess "No Metadata object passed to RightsPredictor" unless $args{'record'};
+  return $self;
+}
+
+# Last year the work was/will be in copyright in country of publication.
+# Returns the following structure:
+#   my $ldc = {
+#     'year' => 1933, # 4-digit int or undef on error
+#     'desc' => 'how derived'
+#   };
+# or undef on error or not enough info.
+sub last_domestic_copyright
+{
+  my $self     = shift;
+  my $year     = shift || ''; # ADD or Pub entered by user
+  my $is_crown = shift; # Crown copyright founds
+
+  if ($year !~ m/^-?\d{1,4}$/) {
+    return { year => undef, desc => "unsupported date format '$year'" };
+  }
+  my $country = $self->{record}->country;
+  if ($country eq 'United Kingdom' || $country eq 'Canada' || $country eq 'Australia') {
+    my $term = 70;
+    if ($is_crown) {
+      $term = 50;
+    } else {
+      if ($country eq 'Canada') {
+        $term = 50 if $year < 1972;
+      } elsif ($country eq 'Australia') {
+        $term = 50 if $year < 1955;
+      }
+    }
+    my $last_domestic_copyright = $year + $term;
+    return { year => $last_domestic_copyright,
+      desc => "Last $country © $last_domestic_copyright ($year + $term-year term)" };
+  }
+  #elsif ($country eq 'Spain') {
+  #  return { year => $year + 80, desc => "$country $year + 80-year term" };
+  #}
+  return { year => undef, desc => "country $country unknown term" };
+}
+
+# CRMS World specific. Predict best radio button (rights combo)
+# to choose based on user selections.
+# Returns data structure with the following fields (all but desc may be absent on error):
+# attr: rights attribute name
+# reason: rights reason name
+# rights: attr/reason string
+# last_domestic_copyright: year
+# desc: last_domestic_copyright desc plus more math, or error message
+sub rights {
+  my $self     = shift;
+  my $year     = shift || ''; # ADD or pub date entered by user
+  my $is_pub   = shift; # Treat $year as pub date for corporate authors
+  my $is_crown = shift; # Crown copyright found
+  #my $pub_year = shift; # Actual pub date when date range, entered by user
+  my $now      = shift || POSIX::strftime("%Y", localtime); # The "current" year
+
+  my $ldc = $self->last_domestic_copyright($year, $is_crown);
```

Comment:
The names in lines 86,87,88 make me have to re-read these lines several times to parse what is happening here.

Reply:
Keeping `$last_source_copyright` as a struct and using the expression `$last_source_copyright->{year}` where needed in the hopes that it will scan better.